### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by adamgraves-choices**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+## Deployment
+
+Deployment is driven by Travis (Dev and PR review deployments) and TeamCity (staging and production deployments). We use the NHSuk [CI-Deployment] repo for deployment to the Rancher environment, this sets the following ENV variables.
+
+## Env vars
+
+| Variable | Description | Default | 
+| -------- | ----------- | ------- |
+| `ES_MIN_MASTER_NODES` | The minimum nodes required to run the host  | `2` |
```
Comment (beseven) [inline README.md]: You mention de default to be 2, and lower down to be 3. Which one is it?

Comment (adamgraves-choices) [reply README.md]: @beseven That's correct.

See the documentation here: https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html#split-brain

The aim is that there's 3 *eligible* master nodes, and this ENV variable is for the *minimum* number of nodes needed.

Comment (st3v3nhunt) [inline README.md]: Should this be `...to run the cluster` rather than `...to run the host`?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# ElasticSearch on Rancher
+
+Configuration for deploying the following services to [Rancher]:
+
+- `es-master`, ElasticSearch cluster, running on every host with a `escluster` tag defined.
```
Comment (st3v3nhunt) [inline README.md]: Would it read better as
> `es-master` node, part of the Elasticsearch cluster, runs on every host that has an `escluster` tag defined

File: README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# ElasticSearch on Rancher
+
```
Comment (st3v3nhunt) [inline README.md:2]: Its good practice to have a short summary of the repo at the top. See https://github.com/RichardLitt/standard-readme/blob/master/spec.md for more info.

If you don't want to read that, just add a sentence or two prepended with an `>` so the text is rendered as quoted.

**Review by st3v3nhunt — DISMISSED** [review]
A few suggestions.

**Comment by adamgraves-choices** [comment]
@st3v3nhunt updated using your notes.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+> Rancher configuration for running a Elasticsearch cluster
+
+## Services deployed:
+
+- `es-master`, Elasticsearch nodes, part of the Elasticsearch cluster, one runs on every host that has an `escluster=true` tag defined.
+- Justwatch [Exporter], exposing Prometheus metrics on the cluster status.
+- es-client, Rancher cattle DNS service, which points the `es-client` domain name to `es-master`, for backwards compatibility with existing services.
+
```
Comment (markysoft) [inline README.md]: Would prefer a colon or a hyphen after the first word,  as it reads a little strange with a comma. something like
`es-master`: Elasticsearch nodes. Part of the Elasticsearch cluster, one runs on every host that has an `escluster=true` tag defined.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# ElasticSearch on Rancher
+
+> Rancher configuration for running a Elasticsearch cluster
+
```
Comment (markysoft) [inline README.md]: should be 'an Elasticsearch...'

File: README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+## Services deployed:
+
+- `es-master`, Elasticsearch nodes, part of the Elasticsearch cluster, one runs on every host that has an `escluster=true` tag defined.
+- Justwatch [Exporter], exposing Prometheus metrics on the cluster status.
+- es-client, Rancher cattle DNS service, which points the `es-client` domain name to `es-master`, for backwards compatibility with existing services.
+
+## Deployment
+
+Deployment is driven by Travis (Dev and PR review deployments) and TeamCity (staging and production deployments). We use the NHSuk [CI-Deployment] repo for deployment to the Rancher environment, this sets the following ENV variables.
```
Comment (markysoft) [inline README.md]: Should `NHSuk` be NHS UK?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+## Env vars
```
Comment (markysoft) [inline README.md]: Other readmes have this as the full words 'Environment variables'

**Review by markysoft — CHANGES_REQUESTED** [review]
Some minor text changes

**Review by markysoft — DISMISSED** [review]

File: README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# ElasticSearch on Rancher
+
+> Rancher configuration for running an Elasticsearch cluster
+
+## Services deployed:
+
```
Comment (neilbmclaughlin) [inline README.md:6]: The rancher-config file has an additional service (`es-sysctl`) not mentioned here.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+- `es-master`: Elasticsearch nodes, part of the Elasticsearch cluster, one runs on every host that has an `escluster=true` tag defined.
+- Justwatch [Exporter]: Exposes Prometheus metrics on the cluster status.
```
Comment (neilbmclaughlin) [inline README.md]: Mixed naming convention `es-master` vs es-client vs Justwatch [Exporter]

My preference would be `es-master`, `es-client` and `es-exporter`

File: README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+- es-client: Rancher cattle DNS service, which points the `es-client` domain name to `es-master`, for backwards compatibility with existing services.
```
Comment (neilbmclaughlin) [inline README.md]: Does the service name have to be `es-client` for the redirect to work. If not then a better name would be something like`es-dns-redirect`. I'm not proposing that this is done as part of this PR.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+## Deployment
+
+Deployment is driven by Travis (Dev and PR review deployments) and TeamCity (staging and production deployments). We use the NHS.UK [CI-Deployment] repo for deployment to the Rancher environment, this sets the following ENV variables.
```
Comment (neilbmclaughlin) [inline README.md]: Maybe a link? (e.g. `[Travis](https://travis-ci.org/nhsuk/rancher-elasticsearch-cluster)`) and for TeamCity too.

File: README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+## How it works
+
+The number of containers running is determined by the number of hosts that have the `escluster` flag set in the Rancher environment. The minimum number should be `3`.
+
+When the containers start they do a DNS A record lookup for `es-master` to get the current hosts in the cluster, ElasticSearch uses this information to build the cluster and for communication.
+
+## How to use the cluster
+
+The cluster is currently setup to only be available inside of the Rancher environment, so your application will have to be deployed to that environment also.
```
Comment (neilbmclaughlin) [inline README.md:31]: I had to do some digging to remind myself of the distinction between _stacks_ and _environments_ so might be useful to link to http://rancher.com/docs/rancher/latest/en/environments/

File: README.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+## Deployment
+
+Deployment is driven by Travis (Dev and PR review deployments) and TeamCity (staging and production deployments). We use the NHS.UK [CI-Deployment] repo for deployment to the Rancher environment, this sets the following ENV variables.
+
+## Environment variables
+
+| Variable | Description | Default | 
+| -------- | ----------- | ------- |
+| `ES_MIN_MASTER_NODES` | The minimum nodes required to run the cluster | `2` |
+| `ES_DATA_VOL` | The name of the local volume on the host to use | `esdata` |
```
Comment (neilbmclaughlin) [inline README.md:21]: I recall that there was a rational for specifying a data volume to do with persistence of data - there is some discussion of this in Slack. Could this be captured here somewhere or maybe even in an ADR.

**Review by neilbmclaughlin — COMMENTED** [review]
Reads well. Most of my comments are where more background would have helped my understanding.

File: README.md (hunk 10 above)
Comment (adamgraves-choices) [reply README.md]: The service name can be anything. I would suggest if we're going to change it, to remove the DNS alias entirely, and just use `es-master` instead.

File: README.md (hunk 9 above)
Comment (adamgraves-choices) [reply README.md]: Your correct, and I'm correct in fact! The service is already called `es-exporter`, just the documentation isn't correct. I'll update it.

File: README.md (hunk 8 above)
Comment (adamgraves-choices) [reply README.md:6]: I'll add it.

File: README.md (hunk 11 above)
Comment (adamgraves-choices) [reply README.md]: Happy to provide the Travis link.

File: README.md (hunk 13 above)
Comment (adamgraves-choices) [reply README.md:21]: I think it's implied that there's a need for a data volume here. That probably should be captured in an ADR.

File: README.md (hunk 12 above)
Comment (adamgraves-choices) [reply README.md:31]: I think that is outside the scope of this documentation.

**Comment by adamgraves-choices** [comment]
@neilbmclaughlin Updated using your notes.

**Review by neilbmclaughlin — APPROVED** [review]

**Pull request merged**